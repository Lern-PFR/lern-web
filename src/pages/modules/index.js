import ModuleDetailsPage from './ModuleDetailsPage';

export {
	// eslint-disable-next-line import/prefer-default-export
	ModuleDetailsPage,
};
