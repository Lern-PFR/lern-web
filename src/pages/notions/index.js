import NotionDetailsPage from './NotionDetailsPage';

export {
	// eslint-disable-next-line import/prefer-default-export
	NotionDetailsPage,
};
