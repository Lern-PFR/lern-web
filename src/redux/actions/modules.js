import * as ModulesApi from 'api/modulesApi';

/**
 * @constant
 * @name ActionTypes
 * @description The various action types used to interact with the modules redux state.
 * @type {object}
 *
 * @author Timothée Simon-Franza
 */
export const ActionTypes = {
	FETCH_MODULE_REQUEST: '@MODULES/FETCH_REQUEST',
	FETCH_MODULE_SUCCESS: '@MODULES/FETCH_SUCCESS',
	FETCH_MODULE_FAILURE: '@MODULES/FETCH_FAILURE',

	FETCH_MODULE_LIST_REQUEST: '@MODULES/FETCH_LIST_REQUEST',
	FETCH_MODULE_LIST_SUCCESS: '@MODULES/FETCH_LIST_SUCCESS',
	FETCH_MODULE_LIST_FAILURE: '@MODULES/FETCH_LIST_FAILURE',

	CREATE_MODULE_REQUEST: '@MODULES/CREATE_REQUEST',
	CREATE_MODULE_SUCCESS: '@MODULES/CREATE_SUCCESS',
	CREATE_MODULE_FAILURE: '@MODULES/CREATE_FAILURE',

	UPDATE_MODULE_REQUEST: '@MODULES/UPDATE_REQUEST',
	UPDATE_MODULE_SUCCESS: '@MODULES/UPDATE_SUCCESS',
	UPDATE_MODULE_FAILURE: '@MODULES/UPDATE_FAILURE',

	DELETE_MODULE_REQUEST: '@MODULES/DELETE_REQUEST',
	DELETE_MODULE_SUCCESS: '@MODULES/DELETE_SUCCESS',
	DELETE_MODULE_FAILURE: '@MODULES/DELETE_FAILURE',

	CLEAR_MODULE_LIST: '@MODULES/CLEAR_LIST',
};

// //////////////////////////////////////////////////////// //
// //////////// Specific module fetching actions ////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchModuleRequest
 * @description Action triggered anytime a module fetching call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const fetchModuleRequest = () => ({ type: ActionTypes.FETCH_MODULE_REQUEST });

/**
 * @function
 * @name fetchModuleSuccess
 * @description Action triggered as a result to a successful module fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} module : The module retrieved from the API.
 *
 * @returns {object}
 */
const fetchModuleSuccess = ({ module }) => ({
	type: ActionTypes.FETCH_MODULE_SUCCESS,
	payload: { module },
});

/**
 * @function
 * @name fetchModuleFailure
 * @description Action triggered as a result to a failed module fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const fetchModuleFailure = (error) => ({
	type: ActionTypes.FETCH_MODULE_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// ////////////// Module list fetching actions //////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchModuleListRequest
 * @description Action triggered anytime a module list fetching call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const fetchModuleListRequest = () => ({ type: ActionTypes.FETCH_MODULE_LIST_REQUEST });

/**
 * @function
 * @name fetchModuleListSuccess
 * @description Action triggered as a result to a successful module list fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {array}	modules		: The list of modules retrieved from the API.
 * @param {number}	totalCount	: The total amount of modules available for the subject.
 *
 * @returns {object}
 */
const fetchModuleListSuccess = ({ modules, totalCount }) => ({
	type: ActionTypes.FETCH_MODULE_LIST_SUCCESS,
	payload: { modules, totalCount },
});

/**
 * @function
 * @name fetchModuleListFailure
 * @description Action triggered as a result to a failed module list fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const fetchModuleListFailure = (error) => ({
	type: ActionTypes.FETCH_MODULE_LIST_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// //////////////// Module creation actions /////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name createModuleRequest
 * @description Action triggered anytime a module creation call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const createModuleRequest = () => ({ type: ActionTypes.CREATE_MODULE_REQUEST });

/**
 * @function
 * @name createModuleSuccess
 * @description Action triggered as a result to a successfult module creation API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} module : The newly created module.
 *
 * @returns {object}
 */
const createModuleSuccess = ({ module }) => ({
	type: ActionTypes.CREATE_MODULE_SUCCESS,
	payload: { module },
});

/**
 * @function
 * @name createModuleFailure
 * @description Action triggered as a result to a failed module creation API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const createModuleFailure = (error) => ({
	type: ActionTypes.CREATE_MODULE_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// //////////////// Module edition actions //////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name updateModuleRequest
 * @description Action triggered anytime a module edition call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const updateModuleRequest = () => ({ type: ActionTypes.UPDATE_MODULE_REQUEST });

/**
 * @function
 * @name updateModuleSuccess
 * @description Action triggered as a result to a successful module edition API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} module : The updated module object.
 *
 * @returns {object}
 */
const updateModuleSuccess = ({ module }) => ({
	type: ActionTypes.UPDATE_MODULE_SUCCESS,
	payload: { module },
});

/**
 * @function
 * @name updateModuleFailure
 * @description Action triggered as a result to a failed module edition API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const updateModuleFailure = (error) => ({
	type: ActionTypes.UPDATE_MODULE_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// //////////////// Module deletion actions //////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name deleteModuleRequest
 * @description Action triggered anytime a module deletion call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const deleteModuleRequest = () => ({ type: ActionTypes.DELETE_MODULE_REQUEST });

/**
 * @function
 * @name deleteModuleSuccess
 * @description Action triggered as a result to a successful module deletion API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} module : The deleted module object.
 *
 * @returns {object}
 */
const deleteModuleSuccess = ({ module }) => ({
	type: ActionTypes.DELETE_MODULE_SUCCESS,
	payload: { module },
});

/**
 * @function
 * @name deleteModuleFailure
 * @description Action triggered as a result to a failed module deletion API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const deleteModuleFailure = (error) => ({
	type: ActionTypes.DELETE_MODULE_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// //////////////// Exported action creators ////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchModule
 * @description Method used to fetch the latest version of a specified module
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} moduleId : The id of the module we want to retrieve.
 */
export const fetchModule = (moduleId) => (dispatch) => {
	dispatch(fetchModuleRequest());

	return ModulesApi.fetchModuleById(moduleId)
		.then(({ module }) => dispatch(fetchModuleSuccess({ module })))
		.catch((error) => dispatch(fetchModuleFailure(error)));
};

/**
 * @function
 * @name fetchModuleListBySubjectId
 * @description Method used to fetch all modules from a specific subject, identified by the subjectId param.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} subjectId : The id of the subject we want to retrieve modules from.
 */
export const fetchModuleListBySubjectId = (subjectId) => (dispatch) => {
	dispatch(fetchModuleListRequest());

	return ModulesApi.fetchModulesBySubjectId(subjectId)
		.then(({ modules, totalCount }) => dispatch(fetchModuleListSuccess({ modules, totalCount })))
		.catch((error) => dispatch(fetchModuleListFailure(error)));
};

/**
 * @function
 * @name createModule
 * @description Method used to create a new module in the database.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} moduleData : The data to create the new module from.
 */
export const createModule = (moduleData) => (dispatch) => {
	dispatch(createModuleRequest());

	return ModulesApi.createModule(moduleData)
		.then(({ module }) => dispatch(createModuleSuccess({ module })))
		.catch((error) => dispatch(createModuleFailure(error)));
};

/**
 * @function
 * @name updateModule
 * @description Method used to update an existing module instance from the database.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} moduleData	: The data to update the module with.
 * @param {string} moduleId		: The id to identify the module to update.
 */
export const updateModule = (moduleData, moduleId) => (dispatch) => {
	dispatch(updateModuleRequest());

	return ModulesApi.updateModule(moduleData, moduleId)
		.then(({ module }) => dispatch(updateModuleSuccess({ module })))
		.catch((error) => dispatch(updateModuleFailure(error)));
};

/**
 * @function
 * @name deleteModule
 * @description Method used to remove an existing module from the database.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} moduleId : The id identifying the module to remove.
 */
export const deleteModule = (moduleId) => (dispatch) => {
	dispatch(deleteModuleRequest());

	return ModulesApi.deleteModule(moduleId)
		.then(({ module }) => dispatch(deleteModuleSuccess({ module })))
		.catch((error) => dispatch(deleteModuleFailure(error)));
};
