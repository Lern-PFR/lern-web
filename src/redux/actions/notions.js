import * as NotionsApi from 'api/notionsApi';

/**
 * @constant
 * @name ActionTypes
 * @description The various action types used to interact with the notions redux state.
 * @type {object}
 *
 * @author Timothée Simon-Franza
 */
export const ActionTypes = {
	FETCH_NOTION_REQUEST: '@NOTIONS/FETCH_REQUEST',
	FETCH_NOTION_SUCCESS: '@NOTIONS/FETCH_SUCCESS',
	FETCH_NOTION_FAILURE: '@NOTIONS/FETCH_FAILURE',

	FETCH_NOTION_LIST_REQUEST: '@NOTIONS/FETCH_LIST_REQUEST',
	FETCH_NOTION_LIST_SUCCESS: '@NOTIONS/FETCH_LIST_SUCCESS',
	FETCH_NOTION_LIST_FAILURE: '@NOTIONS/FETCH_LIST_FAILURE',

	CREATE_NOTION_REQUEST: '@NOTIONS/CREATE_REQUEST',
	CREATE_NOTION_SUCCESS: '@NOTIONS/CREATE_SUCCESS',
	CREATE_NOTION_FAILURE: '@NOTIONS/CREATE_FAILURE',

	UPDATE_NOTION_REQUEST: '@NOTIONS/UPDATE_REQUEST',
	UPDATE_NOTION_SUCCESS: '@NOTIONS/UPDATE_SUCCESS',
	UPDATE_NOTION_FAILURE: '@NOTIONS/UPDATE_FAILURE',

	DELETE_NOTION_REQUEST: '@NOTIONS/DELETE_REQUEST',
	DELETE_NOTION_SUCCESS: '@NOTIONS/DELETE_SUCCESS',
	DELETE_NOTION_FAILURE: '@NOTIONS/DELETE_FAILURE',

	CLEAR_NOTION_LIST: '@NOTIONS/CLEAR_LIST',
};

// //////////////////////////////////////////////////////// //
// //////////// Specific notion fetching actions ////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchNotionRequest
 * @description Action triggered anytime a notion fetching call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const fetchNotionRequest = () => ({ type: ActionTypes.FETCH_NOTION_REQUEST });

/**
 * @function
 * @name fetchNotionSuccess
 * @description Action triggered as a result to a successful notion fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} notion : The notion retrieved from the API.
 *
 * @returns {object}
 */
const fetchNotionSuccess = ({ notion }) => ({
	type: ActionTypes.FETCH_NOTION_SUCCESS,
	payload: { notion },
});

/**
 * @function
 * @name fetchNotionFailure
 * @description Action triggered as a result to a failed notion fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const fetchNotionFailure = (error) => ({
	type: ActionTypes.FETCH_NOTION_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// ////////////// Notion list fetching actions //////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchNotionListRequest
 * @description Action triggered anytime a notion list fetching call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const fetchNotionListRequest = () => ({ type: ActionTypes.FETCH_NOTION_LIST_REQUEST });

/**
 * @function
 * @name fetchNotionListSuccess
 * @description Action triggered as a result to a successful notion list fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {array}	notions		: The list of notions retrieved from the API.
 * @param {number}	totalCount	: The total amount of notions available for the module.
 *
 * @returns {object}
 */
const fetchNotionListSuccess = ({ notions, totalCount }) => ({
	type: ActionTypes.FETCH_NOTION_LIST_SUCCESS,
	payload: { notions, totalCount },
});

/**
 * @function
 * @name fetchNotionListFailure
 * @description Action triggered as a result to a failed notion list fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const fetchNotionListFailure = (error) => ({
	type: ActionTypes.FETCH_NOTION_LIST_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// //////////////// Notion creation actions /////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name createNotionRequest
 * @description Action triggered anytime a notion creation call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const createNotionRequest = () => ({ type: ActionTypes.CREATE_NOTION_REQUEST });

/**
 * @function
 * @name createNotionSuccess
 * @description Action triggered as a result to a successfult notion creation API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} notion : The newly created notion.
 *
 * @returns {object}
 */
const createNotionSuccess = ({ notion }) => ({
	type: ActionTypes.CREATE_NOTION_SUCCESS,
	payload: { notion },
});

/**
 * @function
 * @name createNotionFailure
 * @description Action triggered as a result to a failed notion creation API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const createNotionFailure = (error) => ({
	type: ActionTypes.CREATE_NOTION_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// //////////////// Exported action creators ////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchNotion
 * @description Method used to fetch the latest version of a specified notion.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} notionId : The id of the notion we want to retrieve.
 */
export const fetchNotion = (notionId) => (dispatch) => {
	dispatch(fetchNotionRequest());

	return NotionsApi.fetchNotionById(notionId)
		.then(({ notion }) => dispatch(fetchNotionSuccess({ notion })))
		.catch((error) => dispatch(fetchNotionFailure(error)));
};

/**
 * @function
 * @name fetchNotionListByModuleId
 * @description Method used to fetch all notions from a specific module, identified by the moduleId param.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} moduleId : The id of the module we want to retrieve notions from.
 */
export const fetchNotionListByModuleId = (moduleId) => (dispatch) => {
	dispatch(fetchNotionListRequest());

	return NotionsApi.fetchNotionsByModuleId(moduleId)
		.then(({ notions, totalCount }) => dispatch(fetchNotionListSuccess({ notions, totalCount })))
		.catch((error) => dispatch(fetchNotionListFailure(error)));
};

/**
 * @function
 * @name createNotion
 * @description Method used to create a new notion in the database.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} notionData : The data to create the new notion from.
 */
export const createNotion = (notionData) => (dispatch) => {
	dispatch(createNotionRequest());

	return NotionsApi.createNotion(notionData)
		.then(({ notion }) => dispatch(createNotionSuccess({ notion })))
		.catch((error) => dispatch(createNotionFailure(error)));
};
