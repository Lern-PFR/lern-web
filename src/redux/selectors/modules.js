import { sortBy } from 'lodash';
import { createSelector } from 'reselect';
import { getSubjectById } from './subjects';

/**
 * @function
 * @name getModules
 * @description A selector callback which returns all modules from the Redux state.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} state The redux state at the time of call.
 *
 * @returns {Array}
 */
const getModules = (state) => state?.modules?.items ?? [];

/**
 * @function
 * @name getModuleById
 * @description A selector callback which returns the module whose id matched the moduleId parameter.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} moduleId The id of the module we want to retrieve.
 *
 * @returns {object|undefined}
 */
const getModuleById = createSelector(
	getModules,
	(_, moduleId) => moduleId,
	(modules, moduleId) => {
		const module = modules.filter(({ id }) => (id === moduleId))?.[0] ?? undefined;

		if (module === undefined) {
			return module;
		}

		const concepts = (module.concepts ?? []).map((concept) => {
			const { courses, ...result } = concept; // @TODO: remove this line once the API has been updated.

			const lessons = Object.values((concept.courses || []).reduce((acc, currentLesson) => {
				const currentlyStoredLesson = acc[currentLesson.id];
				if (currentlyStoredLesson) {
					acc[currentLesson.id] = currentlyStoredLesson.version < currentLesson.version ? currentLesson : currentlyStoredLesson;
				} else {
					acc[currentLesson.id] = currentLesson;
				}

				return acc;
			}, {}));

			return {
				...result,
				lessons: sortBy(lessons, 'order'),
			};
		});

		return {
			...module,
			concepts: sortBy(concepts || [], 'order'),
		};
	}
);

/**
 * @function
 * @name getModuleOrderOptions
 * @description A selector callback which returns an array of "order" values for the module edition form's order select field.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {Array}
 */
const getModuleOrderOptions = createSelector(
	getSubjectById,
	(subject) => {
		if (!subject?.modules?.length || subject?.modules?.length < 1) {
			return [{ label: 0, value: 0 }];
		}

		return Array.from({ length: subject?.modules?.length || 0 }, (_, i) => ({ label: i, value: i }));
	}
);

export {
	getModules,
	getModuleById,
	getModuleOrderOptions,
};
