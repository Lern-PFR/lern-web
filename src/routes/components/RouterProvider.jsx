import PropTypes from 'prop-types';
import { Router } from 'react-router-dom';
import { createBrowserHistory } from 'history';

/**
 * @name history
 * @description The global history object, handling the navigation for the whole application
 *
 * @author Yann Hodiesne
 */
export const history = createBrowserHistory();

/**
 * @name RouterProvider
 * @description A component providing a Router from react-router, with the ability to export its history and block navigation
 *
 * @author Yann Hodiesne
 *
 * @param {*} children : The children to render inside the Router
 */
const RouterProvider = ({ children }) => (
	<Router history={history}>
		{children}
	</Router>
);

RouterProvider.propTypes = {
	children: PropTypes.node.isRequired,
};

export default RouterProvider;
