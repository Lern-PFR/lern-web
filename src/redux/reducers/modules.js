import { ActionTypes } from 'redux/actions/modules';

const initialState = {
	isLoading: false,
	items: [],
	totalCount: 0,
};

export default (state = initialState, { type, payload }) => {
	switch (type) {
		case ActionTypes.FETCH_MODULE_REQUEST:
		case ActionTypes.FETCH_MODULE_LIST_REQUEST:
		case ActionTypes.CREATE_MODULE_REQUEST:
		case ActionTypes.UPDATE_MODULE_REQUEST:
		case ActionTypes.DELETE_MODULE_REQUEST:
			return {
				...state,
				isLoading: true,
			};
		case ActionTypes.FETCH_MODULE_FAILURE:
		case ActionTypes.FETCH_MODULE_LIST_FAILURE:
		case ActionTypes.CREATE_MODULE_FAILURE:
		case ActionTypes.UPDATE_MODULE_FAILURE:
		case ActionTypes.DELETE_MODULE_FAILURE:
			return {
				...state,
				isLoading: false,
			};
		case ActionTypes.CREATE_MODULE_SUCCESS:
		case ActionTypes.UPDATE_MODULE_SUCCESS:
		case ActionTypes.DELETE_MODULE_SUCCESS:
			return {
				...state,
				isLoading: false,
			};
		case ActionTypes.FETCH_MODULE_SUCCESS:
			return {
				...state,
				isLoading: false,
				items: [payload.module],
				totalCount: 1,
			};
		case ActionTypes.FETCH_MODULE_LIST_SUCCESS:
			return {
				...state,
				isLoading: false,
				items: [...payload.modules],
				totalCount: payload.totalCount,
			};
		case ActionTypes.CLEAR_MODULE_LIST:
			return {
				...state,
				items: [],
				totalCount: 0,
			};
		default:
			return state;
	}
};
