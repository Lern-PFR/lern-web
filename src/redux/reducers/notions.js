import { ActionTypes } from 'redux/actions/notions';

const initialState = {
	isLoading: false,
	items: [],
	totalCount: 0,
};

export default (state = initialState, { type, payload }) => {
	switch (type) {
		case ActionTypes.FETCH_NOTION_REQUEST:
		case ActionTypes.FETCH_NOTION_LIST_REQUEST:
		case ActionTypes.CREATE_NOTION_REQUEST:
		case ActionTypes.UPDATE_NOTION_REQUEST:
		case ActionTypes.DELETE_NOTION_REQUEST:
			return {
				...state,
				isLoading: true,
			};
		case ActionTypes.FETCH_NOTION_FAILURE:
		case ActionTypes.FETCH_NOTION_LIST_FAILURE:
		case ActionTypes.CREATE_NOTION_FAILURE:
		case ActionTypes.UPDATE_NOTION_FAILURE:
		case ActionTypes.DELETE_NOTION_FAILURE:
			return {
				...state,
				isLoading: false,
			};
		case ActionTypes.CREATE_NOTION_SUCCESS:
		case ActionTypes.UPDATE_NOTION_SUCCESS:
		case ActionTypes.DELETE_NOTION_SUCCESS:
			return {
				...state,
				isLoading: false,
			};
		case ActionTypes.FETCH_NOTION_SUCCESS:
			return {
				...state,
				isLoading: false,
				items: [payload.notion],
				totalCount: 1,
			};
		case ActionTypes.FETCH_NOTION_LIST_SUCCESS:
			return {
				...state,
				isLoading: false,
				items: [...payload.notions],
				totalCount: payload.totalCount,
			};
		case ActionTypes.CLEAR_NOTION_LIST:
			return {
				...state,
				items: [],
				totalCount: 0,
			};
		default:
			return state;
	}
};
