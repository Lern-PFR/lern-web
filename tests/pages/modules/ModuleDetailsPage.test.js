import { mount, shallow } from 'enzyme';
import { ModuleDetailsPage } from 'pages/modules';
import { ActionTypes as NotionsActionTypes } from 'redux/actions/notions';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import RouterProvider from 'routes/components/RouterProvider';
import thunk from 'redux-thunk';
import { act } from 'react-dom/test-utils';

const mockStore = configureMockStore([thunk]);
const store = mockStore({});

describe('ModuleDetailsPage', () => {
	afterEach(() => {
		jest.restoreAllMocks();
	});

	it('should match previous snapshot', () => {
		const sut = (
			<Provider store={store}>
				<RouterProvider>
					<ModuleDetailsPage
						dispatchFetchModuleList={jest.fn()}
						moduleId="1"
						module={{
							name: 'Test module 1',
							description: 'a test module description',
							subjectId: '1',
						}}
						notions={[]}
					/>
				</RouterProvider>
			</Provider>
		);
		const wrapper = shallow(sut);

		expect(wrapper).toMatchSnapshot();
	});

	it('should call the fetchNotionListByModuleId action creator on mount', () => {
		const ownProps = { match: { params: { id: 1 } } };

		mount(
			<Provider store={store}>
				<RouterProvider>
					<ModuleDetailsPage
						subjectId="1"
						module={{
							name: 'Test module 1',
							description: 'a test module description',
							subjectId: '1',
						}}
						notions={[]}
						{...ownProps}
					/>
				</RouterProvider>
			</Provider>
		);

		expect(store.getActions()).toContainEqual({ type: NotionsActionTypes.FETCH_NOTION_LIST_REQUEST });
	});

	it('should call the clearNotionList action creator on unmount', () => {
		const ownProps = { match: { params: { id: 1 } } };

		const wrapper = mount(
			<Provider store={store}>
				<RouterProvider>
					<ModuleDetailsPage
						subjectId="1"
						module={{
							name: 'Test module 1',
							description: 'a test module description',
							subjectId: '1',
						}}
						notions={[]}
						{...ownProps}
					/>
				</RouterProvider>
			</Provider>
		);

		act(() => {
			wrapper.unmount();
		});

		expect(store.getActions()).toContainEqual({ type: NotionsActionTypes.CLEAR_NOTION_LIST });
	});
});
