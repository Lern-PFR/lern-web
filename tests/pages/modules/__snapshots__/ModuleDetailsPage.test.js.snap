// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleDetailsPage should match previous snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <RouterProvider>
    <Connect(ModuleDetailsPage)
      dispatchFetchModuleList={[MockFunction]}
      module={
        Object {
          "description": "a test module description",
          "name": "Test module 1",
          "subjectId": "1",
        }
      }
      moduleId="1"
      notions={Array []}
      t={[Function]}
    />
  </RouterProvider>
</ContextProvider>
`;
